airflow/api_connexion/endpoints/auth_endpoint.py

```diff
@@ -0,0 +1,158 @@
+    """Get site authentication info"""
+    security_manager = current_app.appbuilder.sm
+    config = current_app.config
+    auth_type = security_manager.auth_type
+    type_mapping = {
+        AUTH_DB: "auth_db",
+        AUTH_LDAP: "auth_ldap",
+        AUTH_OID: "auth_oid",
+        AUTH_OAUTH: "auth_oauth",
+        AUTH_REMOTE_USER: "auth_remote_user",
+    }
+    oauth_providers = config.get("OAUTH_PROVIDERS", None)
+    openid_providers = config.get("OPENID_PROVIDERS", None)
+    return info_schema.dump(
+        {
+            "auth_type": type_mapping[auth_type],
+            "oauth_providers": oauth_providers,
+            "openid_providers": openid_providers,
+        }
+    )
+
+
+def auth_login():
+    """Handle DB login"""
+    body = request.json
+    try:
+        data = login_form_schema.load(body)
+    except ValidationError as err:
+        raise Unauthenticated(detail=str(err.messages))
+    security_manager = current_app.appbuilder.sm
+    user = security_manager.login_with_user_pass(data['username'], data['password'])
+    if not user:
+        raise Unauthenticated(detail="Invalid login")
+    login_user(user, remember=False)
+    return security_manager.create_tokens_and_dump(user)
+
+
+def auth_oauthlogin(provider, register=None, redirect_url=None):
+    """Returns OAUTH authorization url"""
+    appbuilder = current_app.appbuilder
+    if register:
+        c_session["register"] = True
+    return appbuilder.sm.oauth_authorization_url(appbuilder.app, provider, redirect_url)
+
+
+def authorize_oauth(provider, state):
+    """Callback to authorize Oauth."""
+    appbuilder = current_app.appbuilder
+    user = appbuilder.sm.oauth_login_user(appbuilder.app, provider, state)
+    return appbuilder.sm.create_tokens_and_dump(user)
+
+
+def auth_remoteuser():
+    """Handle remote user auth"""
+    appbuilder = current_app.appbuilder
+    username = request.environ.get("REMOTE_USER")
+    if username:
+        user = appbuilder.sm.login_remote_user(username)
+    else:
+        raise Unauthenticated(detail="Invalid login")
+    return appbuilder.sm.create_tokens_and_dump(user)
+
+
+@jwt_refresh_token_required
+def refresh_token():
+    """Refresh token"""
+    user = get_jwt_identity()
+    access_token = create_access_token(identity=user)
+    ret = {'access_token': access_token}
+    return jsonify(ret), 200
+
+
+@jwt_required
+def revoke_token():
+    """
+    An endpoint for revoking both access and refresh token.
+
+    This is intended for a case where a logged in user want to revoke
+    another user's tokens
+    """
+    resp = jsonify({"revoked": True})
+    body = request.json
+    try:
+        data = token_schema.load(body)
+    except ValidationError as err:
+        raise BadRequest(detail=str(err.messages))
+    token = decode_token(data['token'])
+    tkn = TokenBlockList.get_token(token['jti'])
+    if not tkn:
+        TokenBlockList.add_token(jti=token['jti'], expiry_delta=token['exp'])
```

Comment:
I would change you you handle this:

Instead of cheacking, and only adding the row if it doesn't exist, instead create the row "blindly" and then catch and ignore the [sqlalchemy.exc.IntegrityError](https://docs.sqlalchemy.org/en/13/core/exceptions.html?highlight=integrity#sqlalchemy.exc.IntegrityError) error.

The reason for doing it this way is that otherwise it is prone to race conditions -- say a user double clicks log out, and this fires of two almost parallel log out requests:

1: logout request starts
2: logout request starts
1: check if token exists - it doesn't
2: check if token exists - it still doesn't
1: add token
2: add token too :BOOM: IntegrityError.

Since it is possible anyway, we'll have to handle it, so we should avoid the extra query to check if the row exists.

Reply:
Nice! Thank you for taking time to explain this!